### PR TITLE
Visualizes blushing on characters after using *blush or when they're drunk

### PR DESCRIPTION
**Pull request by kawoppi**
## About The Pull Request
Using the *blush emote makes your cheeks red for the next 5.2 seconds. The same effect is also visible while you are drunk (while the "drunk" mood event is added).
group2
plblush1
plblush2
(sorry the plasmamen kept burning before i could get a good picture)

## Why It's Good For The Game
It makes the characters a little bit more visually expressive. Also gives the HOP an easier time figuring out who's shitfaced while on the job.

## Changelog

:cl:
expansion: Blushing now makes your face go a little bit red
expansion: Being drunk flushes your cheeks
/:cl:

### CODE REVIEW

**Comment by Onule** [comment]
Could this also be added to drunk characters please?

**Comment by wesoda25** [comment]
Cute

**Comment by thgvr** [comment]
Now THIS is content.

File: code/__DEFINES/traits.dm [hunk 1]
```diff
@@ -359,6 +359,10 @@ Remember to update _globalvars/traits.dm if you're adding/removing/renaming trai
 #define TRAIT_DONT_WRITE_MEMORY "dont_write_memory"
 /// This mob can be painted with the spraycan
 #define TRAIT_SPRAY_PAINTABLE "spray_paintable"
+/// This person is blushing
+#define TRAIT_BLUSHING "blushing"
+/// This person is blushing due to being drunk
+#define TRAIT_DRUNK_BLUSHING "drunk_blushing"
```
Comment (MrMelbert) [inline code/__DEFINES/traits.dm:365]: Traits can actually be assigned and handled from different sources, meaning you don't need both of these.
You only need one trait: `TRAIT_BLUSHING`

File: code/datums/mood_events/drink_events.dm [hunk 2]
```diff
@@ -2,6 +2,16 @@
...
 
+/datum/mood_event/drunk/add_effects(param)
+	// Display blush visual
+	ADD_TRAIT(owner, TRAIT_DRUNK_BLUSHING, SPECIES_TRAIT)
+	owner.update_body()
+
+/datum/mood_event/drunk/remove_effects()
+	// Stop displaying blush visual
+	REMOVE_TRAIT(owner, TRAIT_DRUNK_BLUSHING, SPECIES_TRAIT)
+	owner.update_body()
```
Comment (MrMelbert) [inline code/datums/mood_events/drink_events.dm:13]: The third argument for add and remove trait is the source, and usually it's a descriptor or key related to what added it. 

In this case you're using `SPECIES_TRAIT` which doesn't seem accurate - replace it with something like "DRUNK_TRAIT" or a similar define

File: code/modules/mob/living/emote.dm [hunk 3]
```diff
@@ -4,10 +4,34 @@
...
 	key_third_person = "blushes"
 	message = "blushes."
+	/// Timer for the blush visual to wear off
+	var/blush_timer = TIMER_ID_NULL
+
+/datum/emote/living/blush/run_emote(mob/user, params, type_override, intentional)
+	. = ..()
+	if(. && isliving(user))
+		var/mob/living/living_user = user
+		ADD_TRAIT(living_user, TRAIT_BLUSHING, SPECIES_TRAIT)
```
Comment (MrMelbert) [inline code/modules/mob/living/emote.dm:21]: Same here - instead of applying via "species_trait", use a key like "emote" or something

**Review by MrMelbert — CHANGES_REQUESTED** [review]

**Review by Twaticus — APPROVED** [review]

**Comment by bluetshirtguy** [comment]
the skyrat pipeline...

**Comment by Ryll-Ryll** [comment]
15 seconds feels a little long, no?

File: code/__DEFINES/traits.dm (hunk 1 above)
Comment (kawoppi) [reply code/__DEFINES/traits.dm]: Thanks for the review! I've added EMOTE_TRAIT and DRUNK_TRAIT defines to use as trait sources

**Comment by kawoppi** [comment]
Good thing you brought that up! I set the duration when I was still trying to do other things with it and I didn't think to take another look at it. I've changed it to 5.2 seconds, which seems to match up with how long the text above your head is visible. Although I haven't been able to find the exact duration of it in the code.

**Comment by kawoppi** [comment]
they had something like this already!? maybe this codebase should be a skyrat downstream

File: code/modules/mob/living/carbon/human/species.dm [hunk 4]
```diff
@@ -581,6 +581,14 @@ GLOBAL_LIST_EMPTY(features_by_species)
 					eye_overlay.color = species_human.eye_color
 				standing += eye_overlay
 
+		// blush
+		#define BLUSH_COLOR "#DE5D83"
```
Comment (Mothblocks) [inline code/modules/mob/living/carbon/human/species.dm:585]: Put this in colors instead as like `COLOR_BLUSH_PINK` or similar.

File: code/modules/mob/living/emote.dm [hunk 5]
```diff
@@ -4,10 +4,34 @@
...
+		ADD_TRAIT(living_user, TRAIT_BLUSHING, EMOTE_TRAIT)
+		living_user.update_body()
+
+		// Use a timer to remove the blush effect after the BLUSH_DURATION has passed
+		var/list/key_emotes = GLOB.emote_list["blush"]
+		for(var/datum/emote/living/blush/living_emote in key_emotes)
+			// The existing timer restarts if it's already running
+			blush_timer = addtimer(CALLBACK(living_emote, /datum/emote/living/blush.proc/end_blush, living_user), BLUSH_DURATION, TIMER_UNIQUE | TIMER_OVERRIDE)
+
+/datum/emote/living/blush/proc/end_blush(mob/living/living_user)
```
Comment (Mothblocks) [inline code/modules/mob/living/emote.dm:30]: Might want to check that `living_user` is not `QDELETED`

File: code/modules/mob/living/emote.dm [hunk 6]
```diff
@@ -4,10 +4,34 @@
...
+	if(. && isliving(user))
+		var/mob/living/living_user = user
+		ADD_TRAIT(living_user, TRAIT_BLUSHING, EMOTE_TRAIT)
+		living_user.update_body()
+
+		// Use a timer to remove the blush effect after the BLUSH_DURATION has passed
+		var/list/key_emotes = GLOB.emote_list["blush"]
+		for(var/datum/emote/living/blush/living_emote in key_emotes)
+			// The existing timer restarts if it's already running
+			blush_timer = addtimer(CALLBACK(living_emote, /datum/emote/living/blush.proc/end_blush, living_user), BLUSH_DURATION, TIMER_UNIQUE | TIMER_OVERRIDE)
```
Comment (Mothblocks) [inline code/modules/mob/living/emote.dm:28]:
```suggestion
			blush_timer = addtimer(CALLBACK(living_emote, .proc/end_blush, living_user), BLUSH_DURATION, TIMER_UNIQUE | TIMER_OVERRIDE)
```

File: code/__DEFINES/traits.dm [hunk 7]
```diff
@@ -628,6 +630,8 @@ Remember to update _globalvars/traits.dm if you're adding/removing/renaming trai
 #define MAGIC_TRAIT "magic"
 #define TRAUMA_TRAIT "trauma"
 #define FLIGHTPOTION_TRAIT "flightpotion"
+#define EMOTE_TRAIT "emote"
+#define DRUNK_TRAIT "drunk"
```
Comment (Mothblocks) [inline code/__DEFINES/traits.dm:634]: I guess I won't require it but in the future I always prefer to use `[type]` for this one off stuff.

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/__DEFINES/traits.dm (hunk 7 above)
Comment (kawoppi) [reply code/__DEFINES/traits.dm]: Thanks for the review! I'm not too familiar with this, but I've replaced the trait sources with "[type]" and it looks to be functioning just as before. Hopefully that lines up with how it should be used.

**Pull request merged**